src/backend/app/Services/AgentSoldApplianceService.php

```diff
@@ -151,12 +159,41 @@ public function processSaleFromRequest(AgentSoldAppliance $agentSoldAppliance, a
             'total_cost' => $assignedAppliance->cost,
             'down_payment' => $requestData['down_payment'],
             'asset_id' => $assignedAppliance->appliance->id,
+            'device_serial' => $deviceSerial,
         ];
 
         $appliancePerson = $this->appliancePersonService->make($appliancePersonData);
         $this->agentAppliancePersonService->setAssignee($agent);
         $this->agentAppliancePersonService->setAssigned($appliancePerson);
         $this->agentAppliancePersonService->assign();
+
+        if ($deviceSerial) {
+            $addressFromCustomer = $appliancePerson->person()->first()->addresses()->first();
+            $addressData = $requestData['address'] ?? ['street' => $addressFromCustomer->street, 'city_id' => $addressFromCustomer->city_id];
+            $points = $requestData['points'] ?? $addressFromCustomer->geo()->first()->points;
+            $device = $this->deviceService->getBySerialNumber($deviceSerial);
+            $this->deviceService->update($device, ['person_id' => $requestData['person_id']]);
+            $appliancePerson->device_serial = $deviceSerial;
+            $address = $this->addressesService->make([
+                'street' => $addressData['street'],
+                'city_id' => $addressData['city_id'],
+            ]);
+
+            $this->deviceAddressService->setAssigned($address);
+            $this->deviceAddressService->setAssignee($device);
+            $this->deviceAddressService->assign();
+            $this->addressesService->save($address);
+
+            $geoInfo = $this->geographicalInformationService->make([
+                'points' => $points,
+            ]);
+
+            $this->addressGeographicalInformationService->setAssigned($geoInfo);
+            $this->addressGeographicalInformationService->setAssignee($address);
+            $this->addressGeographicalInformationService->assign();
+            $this->geographicalInformationService->save($geoInfo);
+        }
```

Comment:
For future reference: What this does, is essentially adding the part here, where the device actually get assigned to a customer via `$this->deviceService->update($device, ['person_id' => $requestData['person_id']]);`

This was previously only available to Dashboard sales via the code snippets introduced her: https://github.com/EnAccess/micropowermanager/commit/370ccac18479123726ae1d2926107a1136179d71#diff-97dfbc1b1261522dcc00a657cb567efce666354e4349afecc1cd280345a449ed

But never added for Agent Sales

Reply:
The only slight difference is that it is optional. If the Agent doesn't send a device serial number then we can assume that asset is a non-device asset.